File: src-tauri/src/scanner/export/manifest.rs
use crate::scanner::{CardRecord, ScanResult, ScanStats};
use serde::Serialize;
use std::{
    collections::BTreeMap,
    fs,
    path::{Path, PathBuf},
};

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
struct ManifestIndex<'a> {
    package_root: &'a str,
    streaming_assets: &'a str,
    stats: &'a ScanStats,
    warnings: &'a [String],
    total_cards: usize,
    shards: Vec<ManifestShardInfo>,
}

#[derive(Clone, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
struct ManifestShardInfo {
    key: String,
    game: String,
    href: String,
    card_count: usize,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
struct ManifestShard<'a> {
    key: &'a str,
    game: &'a str,
    cards: Vec<&'a CardRecord>,
}

fn group_cards_by_game(cards: &[CardRecord]) -> BTreeMap<String, Vec<&CardRecord>> {
    let mut grouped = BTreeMap::new();
    for card in cards {
        grouped
            .entry(card.game.clone())
            .or_insert_with(Vec::new)
            .push(card);
    }
    grouped
}

pub(in crate::scanner) fn write_manifest_shards<F>(
    scan: &ScanResult,
    output_root: &Path,
    progress: &mut F,
) -> Result<(PathBuf, usize), String>
where
    F: FnMut(String),
{
    let grouped = group_cards_by_game(&scan.cards);

    let mut shards = Vec::new();
    for (game, cards) in &grouped {
        let key = game.to_ascii_lowercase();
        let href = format!("cards.{key}.json");
        let path = output_root.join(&href);
        let shard = ManifestShard {
            key: &key,
            game,
            cards: cards.clone(),
        };
        progress(format!(
            "Writing manifest shard: {} ({} records)",
            path.display(),
            cards.len()
        ));
        let body = serde_json::to_string(&shard)
            .map_err(|err| format!("Failed to serialize online manifest shard: {err}"))?;
        fs::write(&path, body)
            .map_err(|err| format!("Failed to write online manifest shard: {err}"))?;
        shards.push(ManifestShardInfo {
            key,
            game: game.clone(),
            href,
            card_count: cards.len(),
        });
    }

    let index = ManifestIndex {
        package_root: &scan.package_root,
        streaming_assets: &scan.streaming_assets,
        stats: &scan.stats,
        warnings: &scan.warnings,
        total_cards: scan.cards.len(),
        shards,
    };
    let index_path = output_root.join("cards.index.json");
    progress(format!("Writing manifest index: {}", index_path.display()));
    let body = serde_json::to_string_pretty(&index)
        .map_err(|err| format!("Failed to serialize online manifest index: {err}"))?;
    fs::write(&index_path, body)
        .map_err(|err| format!("Failed to write online manifest index: {err}"))?;

    Ok((index_path, index.shards.len()))
}

#[cfg(test)]
mod tests {
    use super::*;

    fn card(game: &str, id: &str) -> CardRecord {
        CardRecord {
            id: id.to_string(),
            game: game.to_string(),
            record_type: "Card".to_string(),
            data_name: id.to_string(),
            display_name: String::new(),
            character_name: String::new(),
            skill_name: String::new(),
            skill_text: String::new(),
            rare_type: None,
            label_type: None,
            dif_type: None,
            miss: None,
            combo: None,
            chain: None,
            image_path: None,
            thumbnail_path: None,
            asset_layers: Vec::new(),
            source_xml: String::new(),
            editable_fields: Vec::new(),
            print_fields: Vec::new(),
        }
    }

    #[test]
    fn groups_shards_by_game_in_stable_order() {
        let cards = vec![
            card("MU3", "mu3-second"),
            card("CHU", "chu-first"),
            card("MU3", "mu3-first"),
            card("MAI", "mai-first"),
        ];

        let grouped = group_cards_by_game(&cards);
        assert_eq!(
            grouped.keys().map(String::as_str).collect::<Vec<_>>(),
            ["CHU", "MAI", "MU3"]
        );
        assert_eq!(
            grouped["MU3"]
                .iter()
                .map(|card| card.id.as_str())
                .collect::<Vec<_>>(),
            ["mu3-second", "mu3-first"]
        );
    }
}

File: src/runtimeJson/manifest.ts
import type {
  OnlineManifestIndex,
  OnlineManifestShard,
  ScanResult,
  ScanStats,
} from "../types";
import {
  type JsonRecord,
  array,
  finiteNumber,
  invalid,
  nonNegativeInteger,
  record,
  string,
} from "./validation";

function stringArray(value: unknown, source: string, path: string): string[] {
  return array(value, source, path).map((entry, index) =>
    string(entry, source, `${path}[${index}]`),
  );
}

const STAT_KEYS = [
  "chuCards",
  "maiCards",
  "maiCardTypes",
  "maiCardCharas",
  "mu3AssetCards",
  "mu3XmlRecords",
  "pngAssets",
  "unityBundles",
  "unityBundleBytes",
] as const satisfies readonly (keyof ScanStats)[];

function validateStats(value: unknown, source: string, path: string): void {
  const stats = record(value, source, path);
  for (const key of STAT_KEYS) nonNegativeInteger(stats[key], source, `${path}.${key}`);
}

function validateCardRecord(value: unknown, source: string, path: string): void {
  const card = record(value, source, path);
  string(card.id, source, `${path}.id`, false);
  if (card.game !== "CHU" && card.game !== "MAI" && card.game !== "MU3") {
    invalid(source, `${path}.game`, 'one of "CHU", "MAI", or "MU3"');
  }
  for (const key of [
    "recordType",
    "dataName",
    "displayName",
    "characterName",
    "skillName",
    "skillText",
  ] as const) {
    string(card[key], source, `${path}.${key}`);
  }
  for (const key of ["rareType", "labelType", "difType", "miss", "combo", "chain"] as const) {
    if (card[key] !== null) finiteNumber(card[key], source, `${path}.${key}`);
  }
  for (const key of ["imagePath", "thumbnailPath"] as const) {
    if (card[key] !== null) string(card[key], source, `${path}.${key}`);
  }
  if (card.sourceXml !== undefined) string(card.sourceXml, source, `${path}.sourceXml`);

  for (const [index, rawLayer] of array(
    card.assetLayers,
    source,
    `${path}.assetLayers`,
  ).entries()) {
    const layerPath = `${path}.assetLayers[${index}]`;
    const layer = record(rawLayer, source, layerPath);
    string(layer.key, source, `${layerPath}.key`);
    string(layer.label, source, `${layerPath}.label`);
    string(layer.path, source, `${layerPath}.path`, false);
  }
  stringArray(card.editableFields, source, `${path}.editableFields`);
  if (card.editedPrintFields !== undefined) {
    stringArray(card.editedPrintFields, source, `${path}.editedPrintFields`);
  }
  for (const [index, rawField] of array(
    card.printFields,
    source,
    `${path}.printFields`,
  ).entries()) {
    const fieldPath = `${path}.printFields[${index}]`;
    const field = record(rawField, source, fieldPath);
    string(field.key, source, `${fieldPath}.key`);
    string(field.label, source, `${fieldPath}.label`);
    if (
      field.fieldType !== "text" &&
      field.fieldType !== "multiline" &&
      field.fieldType !== "number" &&
      field.fieldType !== "bool" &&
      field.fieldType !== "select" &&
      field.fieldType !== "metadata"
    ) {
      invalid(source, `${fieldPath}.fieldType`, "a supported print field type");
    }
    string(field.value, source, `${fieldPath}.value`);
    if (field.options !== undefined) {
      for (const [optionIndex, rawOption] of array(
        field.options,
        source,
        `${fieldPath}.options`,
      ).entries()) {
        const optionPath = `${fieldPath}.options[${optionIndex}]`;
        const option = record(rawOption, source, optionPath);
        string(option.value, source, `${optionPath}.value`);
        string(option.label, source, `${optionPath}.label`);
      }
    }
  }
}

function validateCards(value: unknown, source: string, path: string): void {
  for (const [index, card] of array(value, source, path).entries()) {
    validateCardRecord(card, source, `${path}[${index}]`);
  }
}

function validateManifestBase(value: JsonRecord, source: string): void {
  string(value.packageRoot, source, "$.packageRoot");
  string(value.streamingAssets, source, "$.streamingAssets");
  validateStats(value.stats, source, "$.stats");
  stringArray(value.warnings, source, "$.warnings");
}

export function parseOnlineManifestIndex(
  value: unknown,
  source = "manifest index",
): OnlineManifestIndex {
  const index = record(value, source, "$");
  validateManifestBase(index, source);
  nonNegativeInteger(index.totalCards, source, "$.totalCards");
  for (const [position, rawShard] of array(index.shards, source, "$.shards").entries()) {
    const path = `$.shards[${position}]`;
    const shard = record(rawShard, source, path);
    string(shard.key, source, `${path}.key`, false);
    string(shard.game, source, `${path}.game`, false);
    string(shard.href, source, `${path}.href`, false);
    nonNegativeInteger(shard.cardCount, source, `${path}.cardCount`);
  }
  return index as OnlineManifestIndex;
}

export function parseOnlineManifestShard(
  value: unknown,
  source = "manifest shard",
): OnlineManifestShard {
  const shard = record(value, source, "$");
  string(shard.key, source, "$.key", false);
  string(shard.game, source, "$.game", false);
  validateCards(shard.cards, source, "$.cards");
  return shard as OnlineManifestShard;
}

export function parseScanResult(value: unknown, source = "legacy manifest"): ScanResult {
  const result = record(value, source, "$");
  validateManifestBase(result, source);
  validateCards(result.cards, source, "$.cards");
  return result as ScanResult;
}
